src/lib/crypto/kimchi_backend/gadgets/keccak.ml

```diff
@@ -0,0 +1,686 @@
+  done ;
+  ()
+
+(*
+* Keccak hash function with input message passed as list of 1byte Cvars.
+* The message will be parsed as follows:
+* - the first byte of the message will be the least significant byte of the first word of the state (A[0][0])
+* - the 10*1 pad will take place after the message, until reaching the bit length rate.
+* - then, {0} pad will take place to finish the 1600 bits of the state.
+*)
+let hash (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (message : Circuit.Field.t list) ~(length : int) ~(capacity : int)
+    (nist : bool) : Circuit.Field.t list =
+  assert (capacity > 0) ;
+  assert (length > 0) ;
+  assert (length mod 8 = 0) ;
+  (* Check each cvar input is 8 bits at most *)
+  check_bytes (module Circuit) message ;
+  let rate = keccak_state_length - capacity in
+  let padded =
+    match nist with
+    | true ->
+        pad_nist (module Circuit) message rate
+    | false ->
+        pad_101 (module Circuit) message rate
+  in
+  let hash = sponge (module Circuit) padded ~length ~capacity ~rate in
+  check_bytes (module Circuit) hash ;
+  (* Check each cvar output is 8 bits at most *)
+  hash
+
+(* Gagdet for NIST SHA-3 function for output lengths 224/256/384/512 *)
+let nist_sha3 (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (len : int) (message : Circuit.Field.t list) : Circuit.Field.t array =
+  let hash =
+    match len with
+    | 224 ->
+        hash (module Circuit) message ~length:224 ~capacity:448 true
+    | 256 ->
+        hash (module Circuit) message ~length:256 ~capacity:512 true
+    | 384 ->
+        hash (module Circuit) message ~length:384 ~capacity:768 true
+    | 512 ->
+        hash (module Circuit) message ~length:512 ~capacity:1024 true
+    | _ ->
+        assert false
+  in
+  Array.of_list hash
+
+(* Gadget for Keccak hash function for the parameters used in Ethereum *)
+let eth_keccak (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (message : Circuit.Field.t list) : Circuit.Field.t array =
+  Array.of_list @@ hash (module Circuit) message ~length:256 ~capacity:512 false
+
+(* KECCAK GADGET TESTS *)
+
+let%test_unit "keccak gadget" =
+  ( if tests_enabled then
+    let (* Import the gadget test runner *)
+    open Kimchi_gadgets_test_runner in
+    (* Initialize the SRS cache. *)
+    let () =
+      try Kimchi_pasta.Vesta_based_plonk.Keypair.set_urs_info [] with _ -> ()
+    in
+
+    let test_keccak ?cs ~nist ~len message expected =
+      let cs, _proof_keypair, _proof =
+        Runner.generate_and_verify_proof ?cs (fun () ->
+            let open Runner.Impl in
+            assert (String.length message % 2 = 0) ;
+            let message =
+              Array.to_list
+              @@ exists
+                   (Typ.array ~length:(String.length message / 2) Field.typ)
+                   ~compute:(fun () ->
+                     Array.of_list
```

Comment:
It would be good to line it up so everything is either `List` or `Array` in order avoid overhead of swapping back and forth.

Reply:
Yeah but each of them have their own limitations. For example, with array it is cheap to access a certain position and modify its value. But for list, it is cheap to append values, but expensive to access a certain position. And Keccak has both things: you need to access values to compose words, but need the ability to put together when padding. It feels like no matter what path you take, it will have overheads in any case. After reading the code, what do you think makes better sense? Array always?

Reply:
It sounds like you've done the best possible since you're faced with both these requirements and limitations.